framework/scripts/agent_upgrade.py

```diff
@@ -82,68 +79,83 @@ def main():
+                                                            rl_timeout=-1 if args.timeout == None else args.timeout, 
+                                                            use_http=use_http)
+
+    counter = 0
+    
+    sleep(10)
+
+    #check if any agent start the upgrade
+    agents = json.loads(upgrade_command_result.replace("\'", "\""))
+
+    id_agents_upgrading = []
+    for agent in agents:
         if not args.silent:
-            if not args.debug:
-                print("\n{0}... Please wait.".format(upgrade_command_result))
-            else:
-                print(upgrade_command_result)
-
-        counter = 0
-        agent_info_stat = os.stat(agent_info).st_mtime
-
-        while agent_info_stat == os.stat(agent_info).st_mtime and counter < common.agent_info_retries:
+            print("Agent id: {0} | Error code: {1} | Data: {2} | Task id: {3}".format(agent['agent'], agent['error'], agent['data'], agent['task_id']))
+            
+        if agent["error"] == 0:
+            id_agents_upgrading.append(agent["agent"])
+    
+    if id_agents_upgrading:
+        #waiting for upgrade task end
+        while counter < common.agent_info_retries and id_agents_upgrading:
             sleep(common.agent_info_sleep)
+            
+            #request state of upgrading tasks
+            upgrade_result = send_task_upgrade_module(command='upgrade_result', agent_list=id_agents_upgrading, debug=args.debug)
+
+            agents_results = json.loads(upgrade_result.replace("\'", "\""))
+
+            if not args.silent:
+                print("Cheking upgrade status: retrie {}".format(counter))             
+            
+            for ag in agents_results:
+                #check one by one if the agent end
+                if ag["status"] == "Outdated":
+                    if not args.silent:
+                        print("Agent upgrade failed: id: {0} status: {1}.".format(ag["agent"], ag["status"]))
+                    id_agents_upgrading.remove(int(ag["agent"]))
+                if ag["status"] == "Upgraded":
+                    if not args.silent:
+                        print("Agent upgraded: id: {0} status: {1}.".format(ag["agent"], ag["status"]))
+                    id_agents_upgrading.remove(int(ag["agent"]))
+            
             counter = counter + 1
 
-        if agent_info_stat == os.stat(agent_info).st_mtime:
-            raise WazuhException(1716, "Timeout waiting for agent reconnection.")
-
-        sleep(10)
-        upgrade_result = agent.upgrade_result(debug=args.debug)
         if not args.silent:
-            if not args.debug:
-                agent.load_info_from_db()
-                print("Agent upgraded: {0} -> {1}".format(prev_ver, agent.version))
+            if id_agents_upgrading:
+                for a in id_agents_upgrading:
+                    print("Agent not finished id: {0}".format(a))
+                raise WazuhException(1716, "Timeout waiting for agent reconnection.")
```

Comment:
Remove this raise since this code is only for debugging.

Reply:
you are right!